EventFilter/EcalRawToDigi/plugins/alpaka/UnpackPortable.dev.cc

```diff
@@ -0,0 +1,437 @@
+          const uint8_t shift = j * 4;            //each channel has 4 bits
+          const int chStatus = (buffer[i] >> shift) & H_CHSTATUS_MASK;
+          const bool regular = (chStatus == CH_DISABLED || chStatus == CH_SUPPRESS);
+          const bool problematic =
+              (chStatus == CH_TIMEOUT || chStatus == CH_HEADERERR || chStatus == CH_LINKERR ||
+               chStatus == CH_LENGTHERR || chStatus == CH_IFIFOFULL || chStatus == CH_L1AIFIFOFULL);
+          if (!(regular || problematic)) {
+            exp_ttids[nCh] = ch;
+            ++nCh;
+          }
+        }
+      }
+
+      //
+      // print Tower block headers
+      //
+      uint8_t ntccblockwords = isBarrel ? 18 : 36;
+      auto const* tower_blocks_start = buffer + 9 + ntccblockwords + 6;
+      auto const* trailer = buffer + (size / 8 - 1);
+      auto const* current_tower_block = tower_blocks_start;
+      uint8_t iCh = 0;
+      uint8_t next_tower_id = exp_ttids[iCh];
+      while (current_tower_block < trailer && iCh < numbChannels) {
+        auto const w = *current_tower_block;
+        uint8_t ttid = w & TOWER_ID_MASK;
+        uint16_t bxlocal = (w >> TOWER_BX_B) & TOWER_BX_MASK;
+        uint16_t lv1local = (w >> TOWER_L1_B) & TOWER_L1_MASK;
+        uint16_t block_length = (w >> TOWER_LENGTH_B) & TOWER_LENGTH_MASK;
+
+        // fast forward to the next good tower id (in case of recovery from an earlier header corruption)
+        while (exp_ttids[iCh] < next_tower_id) {
+          ++iCh;
+        }
+        ++iCh;
+
+        // check if the tower id in the tower header is the one expected
+        // if not try to find the next good header, point the current_tower_block to it, and extract its tower id
+        // or break if there is none
+        if (ttid != next_tower_id) {
+          next_tower_id = find_next_tower_block(current_tower_block, trailer, bx, lv1);
+          if (next_tower_id < TOWER_ID_MASK) {
+            continue;
+          } else {
+            break;
+          }
+        }
+
+        // prepare for the next iteration
+        next_tower_id = exp_ttids[iCh];
+
+        uint16_t const dccbx = bx & 0xfff;
+        uint16_t const dccl1 = lv1 & 0xfff;
+        // fov>=1 is required to support simulated data for which bx==bxlocal==0
+        if (fov >= 1 && !is_synced_towerblock(dccbx, bxlocal, dccl1, lv1local)) {
+          current_tower_block += block_length;
+          continue;
+        }
+
+        // go through all the channels
+        // get the next channel coordinates
+        uint32_t const nchannels = (block_length - 1) / 3;
+
+        bool bad_block = false;
+        auto& ch_with_bad_block = alpaka::declareSharedVar<uint32_t, __COUNTER__>(acc);
+        auto const threadsPerBlock = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
+        // 1 threads per channel in this block
+        for (uint32_t ich = 0; ich < nchannels; ich += threadsPerBlock) {
```

Comment:
In general, the loop should be broken at each `alpaka::syncBlockThreads(acc)` and a new loop should be started after that.

Reply:
Even if the number of channels is not a multiple of 32 all iterations of the loop should reach the `alpaka::syncBlockThreads(acc)` for this loop. Threads that exceed `nchannels` will not perform operations inside the loop before `alpaka::syncBlockThreads(acc)` and reach the `continue` right after it. The next increment of `ich` will be too big to enter the loop body again. Threads that are still within `nchannels` will unpack data before `syncBlockThreads(acc)` and copy it to the output afterwards but also they will not enter the loop another time after that.
Breaking the loop at `syncBlockThreads(acc)` would require additional shared memory for the unpacked data to be used in the second loop.

Reply:
Naively in
```cpp
for (uint32_t ich = 0; ich < nchannels; ich += threadsPerBlock) {
  ...
  alpaka::syncBlockThreads(acc);
```
, when the `nchannels` is not a multiple of 32, the last threads of the block (from `nchannels` to the next multiple of 32) do not call the `alpaka::syncBlockThreads(acc);`.

Perhaps one way would be to round up the `nchannels` to the next multiple of 32 to be used as the loop upper bound here? (given that the `i_to_access` is calculated below, and the later code compares `i_to_access` to `nchannels`)

Reply:
My understanding of the code is this:
Assuming there is one FED with 33 channels. The first iteration of the loop will have `ich` = 0, then second `ich` = 32, and there will not be a third one because `ich` = 64 > 33.

In the second iteration `i_to_access` is between 32 and 63 and the thread 0 will pass the `if (i_to_access < nchannels && i_to_access < ch_with_bad_block)` and do some unpacking. The other threads go directly to `alpaka::syncBlockThreads(acc);`, so all 32 threads will reach it.

After `syncBlockThreads` only thread 0 will pass `if (i_to_access >= nchannels || i_to_access >= ch_with_bad_block) continue;` while the others go to the next iteration of the loop with does not happen because `ich` = 64 > 33.

Am I getting this wrong? The code at least runs just fine with any number of channels the event contains.

Reply:
Good point. I will add a comment.

Reply:
Would it make sense to rename `i_to_access` to `channel` ?

Reply:
By the way, I was trying to think if the approach used here could be mapped to the `uniform_groups`/`uniform_group_elements` or `independent_group`/`independent_group_elements` introduced in CMSSW_14_0_0_pre2 ...

After reading the loop like 10 times, my understanding is that what is being done here (treat the channels as independent if no errors are found, but abort the operation of the faulty channel and any successive one in case an error is found) is too specific for any generalised approach.

Reply:
I guess one could rename `i_to_access` if this makes the code clearer.

Yes your interpretation is correct. If a data corruption is found in one channel it and all successive ones of the block are skipped. This was introduced to avoid the rare HLT crashes we have observed during 2022 (https://github.com/cms-sw/cmssw/issues/39568).

I guess this kernel is a bit non-standard in the sense that the number of channels to unpack is not known before actually starting to unpack and calculating the block length. In addition the channel index is used to pick a 64 bit word in the raw data where the channel starts. And then there is the possibility to skip channels if a different thread finds a data corruption.

Reply:
Understood.

Yes, I think since `i_to_access` does identify a channel number, renaming it to `channel` would make the code easier to follow.